src/util/focus-trap.ts

```diff
@@ -70,36 +80,42 @@ export class NgbFocusTrap {
     }
   }
 
-  /** Detect if incoming focus event should be prevented or not */
-  private _enforceFocus(event) {
+  /** Prevents any focus to occur outside of the watched element */
+  private _stealFocus(event) {
     const {target} = event;
     if (this._document !== target && this._element !== target && !this._element.contains(target)) {
-      this._ngZone.run(() => {
-        const element = this.focusableElement;
-        if (element) {
-          element.focus();
-          event.stopPropagation();
-        }
-      });
+      this._focusElement(this._focusableElement as HTMLElement);
+    }
+  }
+
+  /** Detects if we focus is reaching boundaries of the watched element, preparing to loop */
+  private _enforceFocus(event) {
+    const {target} = event;
+    const[first, last] = this._focusableBoundaryElements;
+    if ((target === first && this._direction === DIRECTION.BACKWARD) ||
+        (target === last && this._direction === DIRECTION.FORWARD)) {
+      this._focusElement(this._focusableElement as HTMLElement);
```

Comment:
You kind of know exaclty which node to focus here, so maybe no need to query document for the second time

Also `as HTMLElement` looks redundant as types match

Reply:
I will split the condition in 2, and focus accordingly